lib/common/issues/src/dashboard.rs

```diff
@@ -0,0 +1,177 @@
+use std::collections::HashMap;
+use std::sync::{Mutex, MutexGuard, OnceLock};
+
+use serde::{Serialize, Serializer};
+
+use crate::issue::{CodeType, Issue, IssueRecord};
+
+#[derive(Default, Serialize)]
+struct Dashboard {
+    pub issues: HashMap<CodeType, IssueRecord>,
+}
+
+impl Dashboard {
+    /// Activates an issue, returning true if the issue was not active before
+    fn add_issue(&mut self, issue: impl Issue) -> bool {
+        let issue = IssueRecord::from(issue);
+        self.issues.insert(issue.code.clone(), issue).is_none()
+    }
+
+    /// Deactivates an issue by its code, returning true if the issue was active before
+    fn remove_issue<S: AsRef<str>>(&mut self, code: S) -> bool {
+        self.issues.remove(code.as_ref()).is_some()
+    }
+}
+
+fn dashboard<'a>() -> MutexGuard<'a, Dashboard> {
+    static DASHBOARD: OnceLock<Mutex<Dashboard>> = OnceLock::new();
+    match DASHBOARD
+        .get_or_init(|| Mutex::new(Dashboard::default()))
+        // this is a blocking attempt to get the dashboard
+        .lock()
+    {
+        Ok(guard) => guard,
+        Err(e) => {
+            log::warn!(
+                "Mutex to the issue dashboard is poisoned, restarting dashboard: {}",
+                e
+            );
+
+            // clear the dashboard
+            let mut guard = e.into_inner();
+            guard.issues.clear();
+
+            guard
+        }
+    }
+}
+
+/// Submits an issue to the dashboard, returning true if the issue code was not active before
+pub fn submit(issue: impl Issue) -> bool {
+    dashboard().add_issue(issue)
+}
+
+/// Solves an issue by its code, returning true if the issue code was active before
+pub fn solve<S: AsRef<str>>(code: S) -> bool {
+    dashboard().remove_issue(code)
+}
+
+/// Generates serialized report of the issues in the dashboard, using the provided serializer
+pub fn report<S: Serializer>(serializer: S) -> Result<S::Ok, S::Error> {
+    dashboard().serialize(serializer)
+}
+
+pub fn all_issues() -> Vec<IssueRecord> {
+    dashboard().issues.values().cloned().collect()
+}
+
+/// Clears all issues from the dashboard
+pub fn clear() {
+    dashboard().issues.clear();
+}
+
+/// Solves (clears) all issues with the given prefix in their code
+pub fn prefix_solve(prefix: &str) {
+    dashboard()
+        .issues
+        .retain(|code, _issue| !code.starts_with(prefix));
```

Comment:
I am not sure I am a fan of the String matching here.
We would need to have a consistent view of all existing code when adding a new one.

Reply:
I'm not sure I understand what you mean here. Adding a new issue can only be done via `insert` operation. This operation **is** expected to be relatively heavy because of scanning the whole hashmap.

Anyway, I only added this as a utility function without a concrete case of when to use it. It should be fine if we remove it.

Reply:
Sorry for not being clear :)

What I meant is that the idea of doing issue grouping based on String prefix is bothering me a bit.

It means the String values are important because adding a new value without checking the existing ones could create problems where an issue is deleted by mistake

Is it clearer?

Reply:
To avoid any premature implementations, I'll just remove these utility functions until there is an actual need for them.